8254758: Change G1ServiceThread to be task based

kstefanj opened the pull request:
Please review this enhancement to make `G1ServiceThread` task based. This thread has evolved from having a single responsibility to do more things. And there are plans to add more to it. To make this easier and to allow different tasks to run at different intervals this change adds a simple task mechanism to the service thread.

The idea is that the service thread keeps a list of `G1ServiceTask`-objects that are ordered by when they should execute. When no tasks are ready to be run the service thread check how far into the future the next task is scheduled and sleeps/waits for that duration. If a new task is added while the thread is waiting it will be woken up to check for ready tasks. 

There are currently two tasks in the list. They have been extracted from the old service thread into the tasks: `G1PeriodicGCTask` and `G1RemSetSamplingTask`. These are more or less identical to the old code, but in the form of tasks. The only difference is that `G1PeriodicGCTask` no longer need to keep track of the last attempt, since it will be rescheduled with the correct interval. 

To reschedule a task is to update the time when it should be executed next and adding it back to the list. This insertion is O(n) and it should be fine since the expected number of tasks handled by the service thread is low. If a task don't want to be rescheduled it can set a negative timeout and that will cause the task to be dropped from the list. There are currently no such tasks and adding such tasks come with the responsibility of making sure resources are taken care of when the task is no longer active.

I've added a gtest to do some basic VM-testing and apart from a lot of local testing I've run mach5 tier 1-4 a few times with good results.

I have created a few follow up tasks to look at moving the current tasks to more appropriate locations and also updated a few old service thread related tasks with the label: [`gc-g1-service-thread`](https://bugs.openjdk.java.net/issues/?jql=labels%20%3D%20gc-g1-service-thread)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ⏳ (2/2 running) | ⏳ (1/2 running) |
| Test (tier1) | ⏳ (7/9 running) | ⏳ (9/9 running) |    |     | 

### Issue
 * [JDK-8254758](https://bugs.openjdk.java.net/browse/JDK-8254758): Change G1ServiceThread to be task based


### Reviewers
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - Author) ⚠️ Review applies to 59a06b0263cf0dff8d84d4e734224a4f11fb3135
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - Author) ⚠️ Review applies to 59a06b0263cf0dff8d84d4e734224a4f11fb3135
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/734/head:pull/734`
`$ git checkout pull/734`

albertnetymk commented:
1. There's a sentinel node in the task list. I wonder if it's possible to get rid of it. This way, `G1ServiceTask` can be made abstract, and `G1ServiceTask::execute` and `G1ServiceTask::timeout` can be pure and unreachable statically.
2. Currently, the task timeout is a `double`, which says very little on what precision is actually supported. I think an integral type would be better.
3. "If a task don't want to be rescheduled it can set a negative timeout and that will cause the task to be dropped from the list." I wonder if `timeout=0` is allowed, and what that means.

kstefanj commented:
>     1. There's a sentinel node in the task list. I wonder if it's possible to get rid of it. This way, `G1ServiceTask` can be made abstract, and `G1ServiceTask::execute` and `G1ServiceTask::timeout` can be pure and unreachable statically.
> 
I tried this out first, but I like not having to check for `NULL` when adding. Another idea I had to was create a specific `G1SentinelTask`, to allow `execute()` and `timeout()` to be pure virtual. If you like this approach I'm happy to go in that direction. 

>     2. Currently, the task timeout is a `double`, which says very little on what precision is actually supported. I think an integral type would be better.
> 
I agree, I would prefer to avoid double all together but it is a quite nice to use `os::elapsedTime()` for the timings since the logging-framework prints time this way. And since you then have a double as the base value having the timeout using the same is quite nice, but I'll take a look an see if I can make it look nice with integral type. But the scheduled time will still be using a double. 

>     3. "If a task don't want to be rescheduled it can set a negative timeout and that will cause the task to be dropped from the list." I wonder if `timeout=0` is allowed, and what that means.

Timeout 0 basically means run me again instantly unless someone else is up. For example looking forward the uncommit task will reschedule itself with a 0 timeout avoiding putting the service thread to sleep but still making it possible for another task to run between two invocations.

kstefanj commented:
Updated after Alberts input:
- Sentinel task is now a separate class, allowing `G1ServiceTask` to be pure virtual.
- Changed `double G1ServiceTask::timeout()` to `int64_t G1ServiceTask::timeout_ms()`

albertnetymk reviewed (approved):
Thank you for the update regarding 1 and 2. As for `timeout=0`, it's not obvious to me how it will be resolved if there are multiple `timeout=0` tasks. Since there are no such tasks yet, I am fine with the current status. Ideally, please provide some doc on the behavior of `timeout=0` next to `virtual int64_t timeout_ms() = 0;`; sth like what's said for "negative" values.

walulyai commented on src/hotspot/share/gc/g1/g1ServiceThread.hpp:
```diff
@@ -28,43 +28,83 @@
+  // Verify that the list is ordered.
+  void verify_task_list() NOT_DEBUG_RETURN;
+public:
+  G1ServiceTaskList();
+  G1ServiceTask* pop();
+  G1ServiceTask* peek();
+  void add_ordered(G1ServiceTask* task);
+  bool is_empty();
```
I think it is better we call it  a priority queue instead of a list.

kstefanj replied:
Sound good, I'll change it to `G1ServiceTaskQueue` and mention it's a priority queue in the comment.

kstefanj commented:
> Thank you for the update regarding 1 and 2. As for `timeout=0`, it's not obvious to me how it will be resolved if there are multiple `timeout=0` tasks. Since there are no such tasks yet, I am fine with the current status. Ideally, please provide some doc on the behavior of `timeout=0` next to `virtual int64_t timeout_ms() = 0;`; sth like what's said for "negative" values.

I will update the comment. The way it will work is that each task is rescheduled using `os::elapsedTime()` and the timeout. So a task with `timeout=0` would still be scheduled after a task with a time value smaller than what's returned by `os::elapsedTime()`. Unless I'm overseeing something, this should be fine with multiple tasks as well.

kstefanj commented on src/hotspot/share/gc/g1/g1ServiceThread.hpp:
```diff
@@ -28,43 +28,76 @@
-
   void stop_service();
 
+  int64_t sleep_time();
```
Good point, I'll update it.

kstefanj reviewed (commented):
Thanks for your comments Ivan.

kstefanj commented:
Latest commit includes some additional renaming apart from what's been discussed here. Had a chat with Ivan and Albert to come up with a bit better names and also moved the logic to decide if a task should be rescheduled to a separate function on the task, rather than using a special value for the `delay_ms()` (former `timeout_ms()`).

albertnetymk commented on src/hotspot/share/gc/g1/g1ServiceThread.cpp:
```diff
@@ -155,19 +149,259 @@ class G1YoungRemSetSamplingClosure : public HeapRegionClosure {
+  }
+
+  // Reschedule task by updating task time and add back to queue.
+  double delay_ms = task->delay_ms() / 1000.0;
```
The delay is in seconds now, right?

albertnetymk commented on src/hotspot/share/gc/g1/g1ServiceThread.cpp:
```diff
@@ -155,19 +149,259 @@ class G1YoungRemSetSamplingClosure : public HeapRegionClosure {
+  G1RemSetSamplingTask remset_task("Remembered Set Sampling Task");
+  register_task(&remset_task);
+
+  while (!should_terminate()) {
```
`should_terminate()` is called in three places/functions; I wonder it's possible to gather them together. Conceptually, we should check termination condition before yielding to the task or to sleep.
Manually inlining `run_tasks` to this loop actually helps readability, since it's quite short itself.

kstefanj replied:
I agree that it's a bit unfortunate and I can see if I can find some way to restructure the code to improve this. Thinking about it, the recent changes might make it possible to nicely just let each round in the outer loop handle one task and then check if we should sleep, maybe this is what you meant.

albertnetymk replied:
This looks good; thank you.

albertnetymk commented on src/hotspot/share/gc/g1/g1ServiceThread.cpp:
```diff
@@ -155,19 +149,259 @@ class G1YoungRemSetSamplingClosure : public HeapRegionClosure {
+    if (os::supports_vtime()) {
+      _vtime_accum = (os::elapsedVTime() - vtime_start);
+    } else {
+      _vtime_accum = 0.0;
```
Why resetting it to zero? I thought empty else branch is enough here.

kstefanj replied:
I think you are correct but this is pre-existing code that I don't intend to change right now. There is a separate [issue](https://bugs.openjdk.java.net/browse/JDK-8252645) on moving this to the G1RemSetSamplingTask. Back in the day when the thread only did remembered set sampling it made sense to measure it here, it doesn't anymore, but I wanted to leave that out of this change.

albertnetymk replied:
Got it, sure.

tschatzl replied:
Okay, let's postpone this to 8252645.

kstefanj reviewed (commented):
Thanks for all input Thomas.

albertnetymk commented:
Now that the task deadline is an integer, `DBL_MAX` should not be used in this PR.

tschatzl commented on src/hotspot/share/gc/g1/g1ServiceThread.hpp:
```diff
@@ -28,43 +28,91 @@
+  // Returns the time in milliseconds until the next task is due.
+  // Used both to determine if there are tasks ready to run and
+  // how long to sleep when nothing is ready.
+  int64_t time_to_next_task_ms();
```
I recommend returning a positive number here because it only raises questions what this means. The implementation of the method also never returns negative values.

tschatzl commented on src/hotspot/share/gc/g1/g1ServiceThread.hpp:
```diff
@@ -28,43 +28,91 @@
 public:
   G1ServiceThread();
   double vtime_accum() { return _vtime_accum; }
+  void register_task(G1ServiceTask* task, jlong delay = 0);
```
Please add a comment what the delay means, i.e. something like: automatically schedules the task with the given delay.

From what is written here I am actually a bit confused about the difference between `register_task()` and `schedule_task()`. One seems to be to use by *other* threads, and the other one should only be called by a `ServiceTask` (in context of the `ServiceThread`).

Do we actually need to make a difference between these two and/or do both need to be public? At first glance I would not be sure who is allowed to call which and what the difference actually is.
It also seems that if either is called by the wrong thread bad things happen. Maybe it is possible to add some asserts to warn about most unintentional misuse?

tschatzl commented on src/hotspot/share/gc/g1/g1ServiceThread.cpp:
```diff
@@ -155,19 +149,249 @@ class G1YoungRemSetSamplingClosure : public HeapRegionClosure {
+  log_debug(gc, task, start)("G1 Service Thread (%s) (run)", task->name());
+  task->execute();
+
+  double duration = os::elapsedVTime() - start;
```
I think this is confusing wrt to all other similar messages. All other concurrent tasks print wall time from start to end. Also for periodicity control (i.e. did it start and end on time) wall time is typically more interesting.

While it is interesting to know how much cpu resources a task took (and for noticing when your machine is overloaded), it is much more interesting to know when it started and ended in absolute terms to see that it was on time or not. I understand that you can derive this yourselves, but this seems tiresome.

Maybe print both similar to "time" output showing both real time and cpu time? I.e. something like

"G1 Service Thread (%s) %1.3fms (cpu: %1.3fms)", task->name(), ...."

(fwiw I think the existing use of elapsedVTime particularly for limiting marking step length in g1 is wrong)

While we are at logging, JFR logging for the tasks would be interesting  for these tasks too. Could you file a new CR for that?

tschatzl commented on src/hotspot/share/gc/g1/g1ServiceThread.cpp:
```diff
@@ -155,19 +132,251 @@ class G1YoungRemSetSamplingClosure : public HeapRegionClosure {
+void G1ServiceTaskQueue::add_ordered(G1ServiceTask* task) {
+  assert(task != NULL, "not a valid task");
+  assert(task->next() == NULL, "invariant");
+  assert(task->time() != DBL_MAX, "invalid time for task");
```
s/DBL_MAX/jlong_max ?

tschatzl commented on src/hotspot/share/gc/g1/g1ServiceThread.hpp:
```diff
@@ -28,43 +28,91 @@
+  virtual void execute();
+};
+
+class G1ServiceTaskQueue {
```
Maybe add:
// Time ordered list of tasks to execute.

kstefanj reviewed (commented):
Thanks for your comments Thomas and Albert.

tschatzl reviewed (approved):
Thanks for your changes. Lgtm.

kstefanj commented:
Thanks @albertnetymk, @tschatzl and @walulyai for the reviews.

kstefanj commented:
/integrate
